### PR TITLE
Add kwargs support to torch.export() API

### LINKED ISSUE #1997: Fix support **kwargs for torch._dynamo.export
### 🐛 Describe the bug

Although `torch._dynamo`'s `export(f, *args, aten_graph=False, decomposition_table=None, tracing_mode='real', **kwargs)` does allow `**kwargs` on its public API, the following snippet

```python
dynamo.config.verbose = True
# torch._dynamo.config.suppress_errors = True
# dynamo.config.dynamic_shapes = False
dynamo.config.log_level = logging.DEBUG
aten_graph = False

def fn_with_kwargs(**kwargs):
    input0 = kwargs["input0"]
    input1 = kwargs["input1"]
    output0 = input0 * input1
    return output0

kwargs = {"input0": torch.randn(4),
            "input1": torch.randn(4)}

graph, _ = torch._dynamo.export(fn_with_kwargs,
                                backend="eager",
                                aten_graph=aten_graph,
                                **kwargs)
result_true = graph(**kwargs)
```

Looking `export`'s implementation, there is a TODO about limitations on `kwargs` flattening

```python
    # TODO(voz): Handle kwargs properly?
    flat_args, in_spec = pytree.tree_flatten(args)
```

which is used to flatten the input to `produce_matching`, which  started the traceback above.

ps: I would love to contrib this fix myself, but opted to create an issue to collect feedback/ideas

### Error logs

```bash
[2022-12-15 21:37:00,697] torch._dynamo.eval_frame: [DEBUG] skipping __init__ /opt/conda/lib/python3.9/contextlib.py
[2022-12-15 21:37:00,697] torch._dynamo.eval_frame: [DEBUG] skipping __enter__ /opt/conda/lib/python3.9/contextlib.py
[2022-12-15 21:37:00,697] torch._dynamo.eval_frame: [DEBUG] skipping __init__ /opt/conda/lib/python3.9/contextlib.py
[2022-12-15 21:37:00,697] torch._dynamo.eval_frame: [DEBUG] skipping __enter__ /opt/conda/lib/python3.9/contextlib.py
[2022-12-15 21:37:00,697] torch._dynamo.eval_frame: [DEBUG] skipping enable_dynamic /opt/conda/lib/python3.9/site-packages/torch/_dynamo/eval_frame.py
[2022-12-15 21:37:00,723] torch._dynamo.symbolic_convert: [INFO] Step 1: torchdynamo start tracing fn_with_kwargs
[2022-12-15 21:37:00,724] torch._dynamo.symbolic_convert: [DEBUG] TRACE starts_line /workspace/dynamo_kwargs.py:38
[2022-12-15 21:37:00,724] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_FAST kwargs []
[2022-12-15 21:37:00,724] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_CONST input0 [ConstDictVariable()]
[2022-12-15 21:37:00,725] torch._dynamo.symbolic_convert: [DEBUG] TRACE BINARY_SUBSCR None [ConstDictVariable(), ConstantVariable(str)]
[2022-12-15 21:37:00,726] torch._dynamo.symbolic_convert: [DEBUG] TRACE STORE_FAST input0 [TensorVariable()]
[2022-12-15 21:37:00,726] torch._dynamo.symbolic_convert: [DEBUG] TRACE starts_line /workspace/dynamo_kwargs.py:39
[2022-12-15 21:37:00,726] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_GLOBAL torch []
[2022-12-15 21:37:00,726] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_ATTR randn_like [TorchVariable(<module 'torch' from '/opt/conda/lib/python3.9/site-packages/torch/__init__.py'>)]
[2022-12-15 21:37:00,727] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_FAST input0 [TorchVariable(<built-in method randn_like of type object at 0x7f0b8d786540>)]
[2022-12-15 21:37:00,727] torch._dynamo.symbolic_convert: [DEBUG] TRACE CALL_FUNCTION 1 [TorchVariable(<built-in method randn_like of type object at 0x7f0b8d786540>), TensorVariable()]
[2022-12-15 21:37:00,729] torch._dynamo.symbolic_convert: [DEBUG] TRACE STORE_FAST input1 [TensorVariable()]
[2022-12-15 21:37:00,730] torch._dynamo.symbolic_convert: [DEBUG] TRACE starts_line /workspace/dynamo_kwargs.py:40
[2022-12-15 21:37:00,730] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_FAST input0 []
[2022-12-15 21:37:00,730] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_FAST input1 [TensorVariable()]
[2022-12-15 21:37:00,730] torch._dynamo.symbolic_convert: [DEBUG] TRACE BINARY_MULTIPLY None [TensorVariable(), TensorVariable()]
[2022-12-15 21:37:00,732] torch._dynamo.symbolic_convert: [DEBUG] TRACE STORE_FAST output0 [TensorVariable()]
[2022-12-15 21:37:00,732] torch._dynamo.symbolic_convert: [DEBUG] TRACE starts_line /workspace/dynamo_kwargs.py:41
[2022-12-15 21:37:00,732] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_FAST output0 []
[2022-12-15 21:37:00,732] torch._dynamo.symbolic_convert: [DEBUG] TRACE RETURN_VALUE None [TensorVariable()]
[2022-12-15 21:37:00,732] torch._dynamo.symbolic_convert: [INFO] Step 1: torchdynamo done tracing fn_with_kwargs
[2022-12-15 21:37:00,734] torch._dynamo.output_graph: [INFO] Step 2: calling compiler function dynamo_normalization_capturing_compiler
Traceback (most recent call last):
  File "/opt/conda/lib/python3.9/site-packages/torch/fx/graph_module.py", line 269, in __call__
    return super(self.cls, obj).__call__(*args, **kwargs)  # type: ignore[misc]
  File "/opt/conda/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1482, in _call_impl
    return forward_call(*args, **kwargs)
  File "<eval_with_key>.2", line 5, in forward
    randn_like = torch.randn_like(kwargs_input0_)
TypeError: randn_like(): argument 'input' (position 1) must be Tensor, not list

Call using an FX-traced Module, line 5 of the traced Module's generated forward function:
def forward(self, kwargs_input0_ : torch.Tensor):
    randn_like = torch.randn_like(kwargs_input0_)

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
    mul = kwargs_input0_ * randn_like;  kwargs_input0_ = randn_like = None

    return (mul,)

[2022-12-15 21:37:00,783] torch._dynamo.debug_utils: [WARNING] Compiled Fx GraphModule failed. Creating script to minify the error.
[2022-12-15 21:37:00,785] torch._dynamo.debug_utils: [WARNING] Writing minified repro to /workspace/torchdynamo_debug/run_2022_12_15_21_37_00_785300/minifier/minifier_launcher.py
Traceback (most recent call last):
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/output_graph.py", line 624, in call_user_compiler
    compiled_fn = compiler_fn(gm, self.fake_example_inputs())
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/debug_utils.py", line 894, in debug_wrapper
    run_fwd_maybe_bwd(compiled_gm, example_inputs)
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/debug_utils.py", line 526, in run_fwd_maybe_bwd
    out = gm(args)
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/eval_frame.py", line 600, in result_capturing_wrapper
    graph_captured_result = graph(*graph_inputs)
  File "/opt/conda/lib/python3.9/site-packages/torch/fx/graph_module.py", line 660, in call_wrapped
    return self._wrapped_call(self, *args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/fx/graph_module.py", line 277, in __call__
    raise e.with_traceback(None)
TypeError: randn_like(): argument 'input' (position 1) must be Tensor, not list

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/workspace/dynamo_kwargs.py", line 54, in <module>
    test_kwargs()
  File "/workspace/dynamo_kwargs.py", line 47, in test_kwargs
    graph, _ = torch._dynamo.export(fn_with_kwargs,
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/eval_frame.py", line 616, in export
    result_traced = opt_f(*args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/eval_frame.py", line 211, in _fn
    return fn(*args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/eval_frame.py", line 332, in catch_errors
    return callback(frame, cache_size, hooks)
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/convert_frame.py", line 103, in _fn
    return fn(*args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/utils.py", line 90, in time_wrapper
    r = func(*args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/convert_frame.py", line 339, in _convert_frame_assert
    return _compile(
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/convert_frame.py", line 398, in _compile
    out_code = transform_code_object(code, transform)
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/bytecode_transformation.py", line 341, in transform_code_object
    transformations(instructions, code_options)
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/convert_frame.py", line 385, in transform
    tracer.run()
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/symbolic_convert.py", line 1676, in run
    super().run()
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/symbolic_convert.py", line 528, in run
    and self.step()
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/symbolic_convert.py", line 496, in step
    getattr(self, inst.opname)(inst)
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/symbolic_convert.py", line 1738, in RETURN_VALUE
    self.output.compile_subgraph(self)
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/output_graph.py", line 477, in compile_subgraph
    self.compile_and_call_fx_graph(tx, list(reversed(stack_values)), root)
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/output_graph.py", line 548, in compile_and_call_fx_graph
    compiled_fn = self.call_user_compiler(gm)
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/output_graph.py", line 629, in call_user_compiler
    raise BackendCompilerFailed(self.compiler_fn, e) from e
torch._dynamo.exc.BackendCompilerFailed: dynamo_normalization_capturing_compiler raised TypeError: randn_like(): argument 'input' (position 1) must be Tensor, not list

Minifier script written to /workspace/torchdynamo_debug/run_2022_12_15_21_37_00_785300/minifier/minifier_launcher.py. Run this script to find the smallest traced graph which reproduces this error.


You can suppress this exception and fall back to eager by setting:
    torch._dynamo.config.suppress_errors = True
```

### Minified repro

```python
import os
from math import inf
import torch
from torch import tensor, device
import torch.fx as fx
import functools
import torch._dynamo
from torch._dynamo.debug_utils import run_fwd_maybe_bwd
from torch._dynamo.optimizations.backends import BACKENDS
from torch._dynamo.testing import rand_strided

# REPLACEABLE COMMENT FOR TESTING PURPOSES

args = [((4,), (1,), torch.float32, 'cpu', False)]
args = [rand_strided(sh, st, dt, dev).requires_grad_(rg) for (sh, st, dt, dev, rg) in args]


from torch.nn import *
class Repro(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, kwargs_input0_ : torch.Tensor):
        randn_like = torch.randn_like(kwargs_input0_)
        mul = kwargs_input0_ * randn_like;  kwargs_input0_ = randn_like = None
        return (mul,)


mod = Repro()

# Setup debug minifier compiler
torch._dynamo.debug_utils.MINIFIER_SPAWNED = True
compiler_fn = BACKENDS["dynamo_minifier_backend"]
raise RuntimeError(
    'Compiler name is None - this likely means that a custom compiler '
    'was called by torchdynamo. Please remove this error, import your '
    'custom compiler function, and replace the compiler_name="None" '
    'line below to compiler_name=<my_imported_custom_function>'
)

dynamo_minifier_backend = functools.partial(
    compiler_fn,
    compiler_name="None",
)
opt_mod = torch._dynamo.optimize(dynamo_minifier_backend)(mod)

with torch.cuda.amp.autocast(enabled=False):
    opt_mod(*args)
(base) root@6639b5aac59b:/workspace# 
```

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
Fixes [#1997](https://github.com/pytorch/torchdynamo/issues/1997)

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/fx/graph.py [hunk 1]
```diff
@@ -624,14 +624,22 @@ def process_outputs(self, out: Any) -> Any:
...
+        # but separate *args and **kwargs for `tree_flatten_spec` call
+        count_args = len(self.pytree_info.in_spec.children_specs[0].children_specs)
+        function_args = self.pytree_info.orig_args[:count_args]
+        function_kwargs = self.pytree_info.orig_args[count_args:]
         if len(free_vars) > 0:  # pytree has placeholders in it
+            function_args = ', '.join(function_args)
+            function_kwargs = '{' + ', '.join(f"'{k}':{v}" for k,v in zip(self.pytree_info.in_spec.children_specs[1].context, function_kwargs)) + '}'
             function_definition += f"""
-    {', '.join(free_vars)}, = fx_pytree.tree_flatten_spec([{', '.join(function_args)}], self._in_spec)"""
+    {', '.join(free_vars)}, = fx_pytree.tree_flatten_spec((({function_args},), {function_kwargs}), self._in_spec)"""
```
Comment (jansel) [inline torch/fx/graph.py:642]: Will `({function_args},)` work in the case of zero args? 

```
>>> (,)
  File "<stdin>", line 1
    (,)
     ^
SyntaxError: invalid syntax
```

**Review by jansel — CHANGES_REQUESTED** [review]

**Comment by vadimkantorov** [comment]
> Fixes #1997

Probably wrong issue mentioned...

**Comment by thiagocrepaldi** [comment]
> > Fixes #1997
> 
> Probably wrong issue mentioned...

Indeed, it was 1997 from pytorch/torchdynamo repo, not pytorch/pytorch. Fixed

File: torch/fx/graph.py (hunk 1 above)
Comment (thiagocrepaldi) [reply torch/fx/graph.py:642]: Thanks. No, doesn't work.
Fixed and added tests

File: torch/fx/graph.py [hunk 2]
```diff
@@ -622,16 +622,33 @@ def process_outputs(self, out: Any) -> Any:
...
         if self.pytree_info is None:
             return super().gen_fn_def(free_vars, maybe_return_annotation)
+
         function_args = self.pytree_info.orig_args
         has_orig_self = (function_args[0] == 'self') if len(function_args) > 0 else False
         if has_orig_self:
             free_vars.insert(0, 'self')
         function_definition = super().gen_fn_def(function_args[:], maybe_return_annotation)
+
+        count_args = len(self.pytree_info.in_spec.children_specs[0].children_specs)
```
Comment (thiagocrepaldi) [inline torch/fx/graph.py:644]: `kwargs` handling is conflicting with `make_fx`, probably when `stateless.function_call` is used. 

This is an existing unit test (test_proxy_tensors.py) that fails with those changes

```python
    def test_make_fx_model_fwd_bwd(self):
        class Foo(torch.nn.Module):
            def __init__(self):
                super().__init__()
                self.linear = torch.nn.Linear(5, 5)

            def forward(self, x):
                return self.linear(x).relu()

        model = Foo()

        def f(x, params):
            out = stateless.functional_call(model, params, x).sum()
            out.backward()
            return list(params.values())
        input = torch.randn(3, 5, requires_grad=True)
        params = dict(model.named_parameters())
        fx_f = make_fx(f, tracing_mode=self.tracing_mode)(input, params)
```

Any idea on how to go about this? A naive idea would be to override `_PyTreeCodeGen` and its `gen_fn_def` method when `torch.export` is used, but maybe there is a more general solution

**Review by jansel — APPROVED** [review]

**Comment by DanilBaibak** [comment]
PR breaks the internal build. @jansel could you pls help fix it. Here are you can find more details - [D42528292](https://www.internalfb.com/diff/D42528292).

**Comment by jansel** [comment]
@tugsbayasgalan looks like you wrote the internal test that is failing (test_trace_export (executorch.exir.tests.test_experimental.TestExperimental)) -- can you help explain the issue to @thiagocrepaldi and ideally create a public test for it.

**Comment by thiagocrepaldi** [comment]
> @kit1980 has imported this pull request. If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42652275).

Hi @kit1980 Is the import process finished? If not, should we manually merge this PR again?

File: torch/_dynamo/eval_frame.py [hunk 3]
```diff
@@ -665,9 +664,10 @@ def graph_with_interpreter(*args):
     ).transform()
 
     # Make dynamo graph to have same input/output spec as user code
+    input_strs = [f"orig_arg_{i}" for i in range(len(args))] + list(kwargs.keys())
```
Comment (tugsbayasgalan) [inline torch/_dynamo/eval_frame.py:667]: Should this be conditional check where we only append kwargs list if it is not empty?

**Comment by gmagogsfm** [comment]
> > Is the import process finished? If not, should we manually merge this PR again?
> 
> @thiagocrepaldi I'm seeing tons of executorch errors internally, need to figure that out first.

@tugsbayasgalan  has landed a change to get rid of the errors internally. We can try to land this PR again.

File: torch/_dynamo/eval_frame.py (hunk 3 above)
Comment (thiagocrepaldi) [reply torch/_dynamo/eval_frame.py:667]: @tugsbayasgalan I will merge this PR, but I will come back to discuss your concerns and hopefully address the corner case detected internally

**Comment by thiagocrepaldi** [comment]
@pytorch merge -g

**Comment by thiagocrepaldi** [comment]
@kit1980 could you import this PR? There is a check complaining about "Meta Internal-Only Changes Check — The Diff and Pull Request are not in sync!"

**Comment by kit1980** [comment]
> @kit1980 could you import this PR? There is a check complaining about "Meta Internal-Only Changes Check — The Diff and Pull Request are not in sync!"

@thiagocrepaldi I've imported, but let me verify the fix we have actually works. I'll merge this myself if it works.

**Comment by kit1980** [comment]
@pytochbot rebase -b master

**Pull request closed without merging**